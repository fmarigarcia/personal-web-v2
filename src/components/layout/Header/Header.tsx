import { useSmoothScroll } from '@hooks/useSmoothScroll';
import { Logo } from '@components/ui/Logo';
import { NavigationLinks } from '@components/ui/NavigationLinks';
import { LanguageToggle } from '@components/ui/LanguageToggle';
import { MobileMenuButton } from '@components/ui/MobileMenuButton';

export const Header: React.FC = () => {
    const { scrollToElement } = useSmoothScroll();

    const handleNavClick = (sectionId: string) => {
        scrollToElement(sectionId);
    };

    return (
        <header
            className="fixed top-0 w-full z-50 bg-white/80 backdrop-blur-md border-b border-gray-200"
            role="banner"
        >
            <div className="container mx-auto px-6 py-4">
                <nav
                    className="flex items-center justify-between"
                    role="navigation"
                    aria-label="Main navigation"
                >
                    <Logo onClick={() => handleNavClick('hero')} />

                    <NavigationLinks onNavClick={handleNavClick} />

                    <div className="flex items-center space-x-4">
                        <LanguageToggle />
                        <MobileMenuButton />
                    </div>
                </nav>
            </div>
        </header>
    );
};
