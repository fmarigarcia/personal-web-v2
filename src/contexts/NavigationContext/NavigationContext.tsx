import React, { createContext, useContext, useState } from 'react';
import type { ReactNode } from 'react';
import { SECTIONS } from '@utils/constants';

// Types
interface NavigationContextType {
    currentSection: string;
    setCurrentSection: (section: string) => void;
    isScrolling: boolean;
    setIsScrolling: (scrolling: boolean) => void;
}

// Create context
const NavigationContext = createContext<NavigationContextType | undefined>(
    undefined
);

// Provider component
interface NavigationProviderProps {
    children: ReactNode;
}

export const NavigationProvider: React.FC<NavigationProviderProps> = ({
    children,
}) => {
    const [currentSection, setCurrentSection] = useState<string>(SECTIONS.HERO);
    const [isScrolling, setIsScrolling] = useState<boolean>(false);

    const value: NavigationContextType = {
        currentSection,
        setCurrentSection,
        isScrolling,
        setIsScrolling,
    };

    return (
        <NavigationContext.Provider value={value}>
            {children}
        </NavigationContext.Provider>
    );
};

// Hook to use the context
export const useNavigation = (): NavigationContextType => {
    const context = useContext(NavigationContext);

    if (context === undefined) {
        throw new Error(
            'useNavigation must be used within a NavigationProvider'
        );
    }

    return context;
};
