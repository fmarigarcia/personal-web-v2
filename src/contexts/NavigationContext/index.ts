export {
    NavigationContext,
    type NavigationContextType,
} from './NavigationContext';
export { NavigationProvider } from './NavigationProvider';
export { useNavigation } from './useNavigation';
