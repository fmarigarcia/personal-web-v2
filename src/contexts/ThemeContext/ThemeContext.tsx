import { createContext } from 'react';
//@ts-expect-error random error
import type { ThemeContextType } from '@types/theme';

export const ThemeContext = createContext<ThemeContextType | undefined>(
    undefined
);
