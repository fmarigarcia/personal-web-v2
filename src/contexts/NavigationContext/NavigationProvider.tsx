import React, { useState, useCallback, useRef } from 'react';
import type { ReactNode } from 'react';
import { SECTIONS } from '@utils/constants';
import { NavigationContext } from './NavigationContext';
import type { NavigationContextType } from './NavigationContext';

interface NavigationProviderProps {
    children: ReactNode;
}

export const NavigationProvider: React.FC<NavigationProviderProps> = ({
    children,
}) => {
    const [currentSection, setCurrentSection] = useState<string>(SECTIONS.HERO);
    const [isScrolling, setIsScrolling] = useState<boolean>(false);

    // Use a ref to store the navigation function to avoid infinite loops
    const navigateRef = useRef<((sectionId: string) => void) | undefined>(
        undefined
    );

    // Stable navigation function that calls the current ref value
    const navigateToSection = useCallback((sectionId: string) => {
        if (navigateRef.current) {
            navigateRef.current(sectionId);
        }
    }, []);

    // Stable setter function that updates the ref
    const setNavigateToSection = useCallback(
        (fn: (sectionId: string) => void) => {
            navigateRef.current = fn;
        },
        []
    );

    const value: NavigationContextType = {
        currentSection,
        setCurrentSection,
        isScrolling,
        setIsScrolling,
        navigateToSection,
        setNavigateToSection,
    };

    return (
        <NavigationContext.Provider value={value}>
            {children}
        </NavigationContext.Provider>
    );
};
