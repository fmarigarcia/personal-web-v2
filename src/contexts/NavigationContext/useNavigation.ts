import { useContext } from 'react';
import { NavigationContext } from './NavigationContext';
import type { NavigationContextType } from './NavigationContext';

// Hook to use the context
export const useNavigation = (): NavigationContextType => {
    const context = useContext(NavigationContext);

    if (context === undefined) {
        throw new Error(
            'useNavigation must be used within a NavigationProvider'
        );
    }

    return context;
};
