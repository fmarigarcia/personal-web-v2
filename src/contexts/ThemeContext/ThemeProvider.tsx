import React, { useState, useEffect } from 'react';

import { THEMES } from '@utils/constants';

import { ThemeContext } from './ThemeContext';

import type { Theme } from '../../types/theme';

interface ThemeProviderProps {
    children: React.ReactNode;
}

export const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
    const [theme, setTheme] = useState<Theme>(() => {
        // Check for saved theme in localStorage
        const savedTheme = localStorage.getItem('theme') as Theme;
        if (
            savedTheme &&
            (savedTheme === THEMES.LIGHT || savedTheme === THEMES.DARK)
        ) {
            return savedTheme;
        }

        // Check for system preference
        if (typeof window !== 'undefined') {
            return window.matchMedia('(prefers-color-scheme: dark)').matches
                ? THEMES.DARK
                : THEMES.LIGHT;
        }

        return THEMES.LIGHT;
    });

    useEffect(() => {
        const root = document.documentElement;

        if (theme === THEMES.DARK) {
            root.classList.add('dark');
        } else {
            root.classList.remove('dark');
        }

        localStorage.setItem('theme', theme);
    }, [theme]);

    const toggleTheme = () => {
        setTheme((prevTheme) =>
            prevTheme === THEMES.LIGHT ? THEMES.DARK : THEMES.LIGHT
        );
    };

    const contextValue = {
        theme,
        toggleTheme,
        setTheme,
    };

    return (
        <ThemeContext.Provider value={contextValue}>
            {children}
        </ThemeContext.Provider>
    );
};
