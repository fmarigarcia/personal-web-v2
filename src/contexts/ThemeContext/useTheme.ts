import { useContext } from 'react';
import { ThemeContext } from './ThemeContext';
import type { UseThemeReturn } from '../../types/theme';

export const useTheme = (): UseThemeReturn => {
    const context = useContext(ThemeContext);

    if (!context) {
        throw new Error('useTheme must be used within a ThemeProvider');
    }

    const { theme, toggleTheme, setTheme } = context;

    const data = {
        theme,
        isDark: theme === 'dark',
        isLight: theme === 'light',
    };

    const actions = {
        toggleTheme,
        setTheme,
    };

    return { data, actions };
};
