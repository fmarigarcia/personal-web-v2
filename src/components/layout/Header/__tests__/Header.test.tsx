import { render, screen, fireEvent } from '@testing-library/react';
import { NavigationProvider } from '@contexts/NavigationContext';
import { ThemeProvider } from '@contexts/ThemeContext';
import { Header } from '../Header';

// Mock i18next
jest.mock('react-i18next', () => ({
    useTranslation: () => ({
        t: (key: string) => {
            const translations: Record<string, string> = {
                'nav.home': 'Home',
                'nav.about': 'About',
                'nav.experience': 'Experience',
                'nav.contact': 'Contact',
                'nav.mainNavigation': 'Main navigation',
            };
            return translations[key] || key;
        },
        i18n: {
            changeLanguage: jest.fn(),
            language: 'en',
        },
    }),
}));

// Mock hooks
jest.mock('@hooks/useSmoothScroll', () => ({
    useSmoothScroll: () => ({
        data: {},
        actions: {
            scrollToElement: jest.fn(),
        },
    }),
}));

// Mock platform hook
jest.mock('@hooks/usePlatform', () => ({
    usePlatform: () => ({
        data: {
            isDesktop: true,
            isMobile: false,
            isTablet: false,
        },
    }),
}));

// Mock localStorage
const mockLocalStorage = {
    getItem: jest.fn().mockReturnValue(null),
    setItem: jest.fn(),
};
Object.defineProperty(window, 'localStorage', {
    value: mockLocalStorage,
});

// Mock matchMedia
Object.defineProperty(window, 'matchMedia', {
    value: jest.fn().mockImplementation(() => ({
        matches: false,
        media: '',
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

describe('Header', () => {
    const renderHeader = () => {
        return render(
            <ThemeProvider>
                <NavigationProvider>
                    <Header />
                </NavigationProvider>
            </ThemeProvider>
        );
    };

    it('should render navigation items', () => {
        renderHeader();

        expect(screen.getByText('Home')).toBeInTheDocument();
        expect(screen.getByText('About')).toBeInTheDocument();
        expect(screen.getByText('Experience')).toBeInTheDocument();
        expect(screen.getByText('Contact')).toBeInTheDocument();
    });

    it('should render language toggle buttons', () => {
        renderHeader();

        expect(
            screen.getByRole('button', { name: /switch to english/i })
        ).toBeInTheDocument();
        expect(
            screen.getByRole('button', { name: /switch to spanish/i })
        ).toBeInTheDocument();
    });

    it('should handle navigation click', () => {
        renderHeader();

        const aboutLink = screen.getByText('About');
        fireEvent.click(aboutLink);

        // The click should not cause any errors
        expect(aboutLink).toBeInTheDocument();
    });

    it('should handle language change', () => {
        renderHeader();

        const spanishButton = screen.getByRole('button', {
            name: /switch to spanish/i,
        });
        fireEvent.click(spanishButton);

        // The click should not cause any errors
        expect(spanishButton).toBeInTheDocument();
    });

    it('should have proper accessibility attributes', () => {
        renderHeader();

        const nav = screen.getByRole('navigation');
        expect(nav).toBeInTheDocument();
        expect(nav).toHaveAttribute('aria-label');

        // Check navigation buttons specifically (not theme toggle)
        const navButtons = screen.getAllByRole('button').filter((button) => {
            const text = button.textContent;
            return ['Home', 'About', 'Experience', 'Contact'].includes(
                text || ''
            );
        });

        navButtons.forEach((button) => {
            expect(button).toHaveAttribute('type', 'button');
        });
    });

    it('should match snapshot', () => {
        const { container } = renderHeader();
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should be responsive', () => {
        renderHeader();

        const header = screen.getByRole('banner');
        expect(header).toHaveClass('fixed', 'top-0', 'w-full');
    });

    it('should highlight current section', () => {
        renderHeader();

        // The home/hero section should be active by default
        const homeLink = screen.getByText('Home');
        expect(homeLink.closest('button')).toHaveClass('text-stone-950');
    });
});
