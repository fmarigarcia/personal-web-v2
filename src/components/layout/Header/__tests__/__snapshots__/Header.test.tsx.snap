// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match snapshot 1`] = `
<header
  class="fixed top-0 w-full z-50 bg-white/80 backdrop-blur-md border-b border-gray-200"
  role="banner"
>
  <div
    class="container mx-auto px-6 py-4"
  >
    <nav
      aria-label="Main navigation"
      class="flex items-center justify-between"
      role="navigation"
    >
      <div
        class="flex items-center"
      >
        <button
          aria-label="Go to home"
          class="text-xl font-bold text-gray-900 hover:text-blue-600 transition-colors"
          type="button"
        >
          FM
        </button>
      </div>
      <div
        class="hidden md:flex items-center space-x-8"
      >
        <button
          aria-label="Go to Home section"
          class="text-sm font-medium transition-colors hover:text-blue-600 text-blue-600"
          type="button"
        >
          Home
        </button>
        <button
          aria-label="Go to About section"
          class="text-sm font-medium transition-colors hover:text-blue-600 text-gray-700"
          type="button"
        >
          About
        </button>
        <button
          aria-label="Go to Experience section"
          class="text-sm font-medium transition-colors hover:text-blue-600 text-gray-700"
          type="button"
        >
          Experience
        </button>
        <button
          aria-label="Go to Contact section"
          class="text-sm font-medium transition-colors hover:text-blue-600 text-gray-700"
          type="button"
        >
          Contact
        </button>
      </div>
      <div
        class="flex items-center space-x-4"
      >
        <div
          class="flex items-center space-x-2"
        >
          <button
            aria-label="Switch to English"
            class="text-sm font-medium px-2 py-1 rounded transition-colors bg-blue-600 text-white"
            type="button"
          >
            EN
          </button>
          <button
            aria-label="Switch to Spanish"
            class="text-sm font-medium px-2 py-1 rounded transition-colors text-gray-700 hover:text-blue-600"
            type="button"
          >
            ES
          </button>
        </div>
        <button
          aria-expanded="false"
          aria-label="Open mobile menu"
          class="md:hidden p-2 text-gray-700 hover:text-blue-600 transition-colors"
          type="button"
        >
          <svg
            class="w-6 h-6"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
          >
            <path
              d="M4 6h16M4 12h16M4 18h16"
            />
          </svg>
        </button>
      </div>
    </nav>
  </div>
</header>
`;
